File: Medium/15. 3Sum/3Sum.java
/*
LeetCode Medium: 15. 3Sum
https://leetcode.com/problems/3sum/

Check 3Sum.cpp for brief solution explanation.
*/
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int N = nums.length;
        Arrays.sort(nums);
        List<List<Integer>>ans = new LinkedList<>();
        for(int l = 0;l<N-2;l++){
            int m = l+1, r = N-1;
            while(m<r&&m<N&&r<N){
                int sum = nums[m]+nums[r];
                if(sum<-nums[l]){
                    m++;
                }else if(sum>-nums[l]){
                    r--;
                }else{
                    List<Integer>t = new LinkedList<Integer>();
                    t.add(nums[l]);
                    t.add(nums[m]);
                    t.add(nums[r]);
                    ans.add(t);
                    int mid = nums[m], right = nums[r];
                    while(m<r&&nums[m]==mid){
                        m++;
                    }
                    while(m<r&&nums[r]==right){
                        r--;
                    }
                }
            }
            while(l+1<N-2&&nums[l]==nums[l+1]){
                l++;
            }
        }
        return ans;
    }
}

File: Medium/15. 3Sum/3Sum.py
'''
LeetCode Medium: 15. 3Sum
https://leetcode.com/problems/3sum/

Check 3Sum.cpp for brief solution explanation.
'''
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        N = len(nums)
        nums.sort()
        ans = []
        for l in range(N-2):
            if l>0 and nums[l-1]==nums[l]:
                continue
            m, r = l+1, N-1
            while m<r and m<N and r<N:
                cur_sum = nums[m]+nums[r]
                if cur_sum<-nums[l]:
                    m+=1
                elif cur_sum>-nums[l]:
                    r-=1
                else:
                    t = [nums[l], nums[m], nums[r]]
                    ans.append(t)
                    while m<r and nums[m]==t[1]:
                        m+=1
                    while m<r and nums[r]==t[2]:
                        r-=1
        return ans